fastlane_bot/helpers/txhelpers.py

```diff
@@ -117,578 +53,136 @@ def validate_and_submit_transaction(
-                "jsonrpc": "2.0",
-                "method": "eth_createAccessList",
-                "params": [
-                    {
-                        "from": self.wallet_address,
-                        "to": self.arb_contract.address,
-                        "gas": expected_gas,
-                        "data": transaction_data,
-                    }
-                ],
-            }
-            if eth_input is None
-            else {
-                "id": 1,
-                "jsonrpc": "2.0",
-                "method": "eth_createAccessList",
-                "params": [
-                    {
-                        "from": self.wallet_address,
-                        "to": self.arb_contract.address,
-                        "gas": expected_gas,
-                        "value": hex(eth_input),
-                        "data": transaction_data,
-                    }
-                ],
-            }
-        )
-        response = requests.post(self.alchemy_api_url, json=json_data)
-        if "failed to apply transaction" in response.text:
+        tx, gas_price = self._build_transaction(function=function, value=value)
+        if tx is None:
             return None
-        else:
-            access_list = loads(response.text)["result"]["accessList"]
-            return access_list
 
-    def construct_contract_function(
-        self,
-        routes: List[Dict[str, Any]],
-        src_amt: int,
-        src_address: str,
-        flashloan_struct=None,
-    ):
-        """
-        Builds the transaction using the Arb Contract function. This version can generate transactions using flashloanAndArb and flashloanAndArbV2.
+        if self.use_access_lists:
+            try:
+                result = self.cfg.w3.eth.create_access_list(tx)
+                if tx["gas"] > result["gasUsed"]:
+                    tx["gas"] = result["gasUsed"]
+                    tx["accessList"] = [{"address": item["address"], "storageKeys": [key.hex() for key in item["storageKeys"]]} for item in result["accessList"]]
+            except Exception as e:
+                self.cfg.logger.info(f"create_access_list({tx}) failed with {e}")
 
-        routes: the routes to be used in the transaction
-        src_amt: the amount of the source token to be sent to the transaction
-        gas_price: the gas price to be used in the transaction
+        tx["gas"] += self.cfg.DEFAULT_GAS_SAFETY_OFFSET
 
-        returns: the transaction function ready to be submitted
-        """
-        if self.ConfigObj.SELF_FUND:
-            return self.arb_contract.functions.fundAndArb(
-                routes, src_address, src_amt
-            )
-        return self.arb_contract.functions.flashloanAndArbV2(
-            flashloan_struct, routes
-        )
+        raw_tx = self.cfg.w3.eth.account.sign_transaction(tx, self.cfg.ETH_PRIVATE_KEY_BE_CAREFUL).rawTransaction
 
-    def _handle_function_build_exception(self, exception: Exception) -> int or None:
```

Comment:
Removing `_handle_function_build_exception` is extremely sub-optimal. The exception returns a valid baseFee, letting the bot proceed with what could be a perfectly valid arb opportunity without rerunning the entire cycle. By forcing the cycle to restart we have a high likelihood of losing the opportunity due to the time it takes to rerun the cycle.

Reply:
This error happens when the block's `baseFee` becomes higher than the transaction's `maxGasFee`, which theoretically yields a negative tip to the miner, hence the transaction is immediately reverted. Recall that the miner gets paid `maxGasFee - baseFee`, regardless of the specified `maxPriorityFee`, which indicates the maximum tip and not the actual tip. But even before this happens, the block's `baseFee` may become sufficiently high to bring the effective tip paid to the miner extremely small, thus leading the transaction to linger for a very long time, and most likely become invalid (arbitrage already fulfilled) by the time it gets executed. So the only difference between these two scenarios is a reverted transaction vs a stuck transaction. Both can sporadically happen, just like front-running can sporadically invalidate arbs submitted by this bot, for example. This error can be handled in a "natural manner", at the next iteration of the bot's infinite loop, hence there is no need to tailor-make a special "handle error and retry" method for this specific error.